text/0000-template.md

```diff
@@ -0,0 +1,78 @@
+
+## Motivation
+
+The difference between `{{let}}` and `{{with}}` is with how they handle conditional arguments. The `{{let}}` helper's block content is always rendered, regardless of its parameters. In contrast, `{{with}}` only renders its main block when the first position parameter is truthy. For example:
+
+```hbs
+{{#with "Alex" as |value|}}
+  {{value}} is truthy
+{{else}}
+  The first positional param was falsy
+{{/with}}
+```
+
+Will render "[Alex] is truthy".
+
+```hbs
+{{#with false as |value|}}
+  {{value}} is truthy
+{{else}}
+  The first positional param was falsy
+{{/with}}
+```
+
+Will render "The first positional param was falsy".
+
+The conditional arguments behavior of `{{with}}` can easily be replicated using a combination of `{{let}}` and `{{if}}` in a way that's easily readable:
+
+```hbs
+{{#let "Alex" as |value|}}
+  {{#if value}}
+    {{value}} is truthy
+  {{else}}
+    The first positional param was falsy
+  {{/if}}
+{{/let}}
+```
+
+## Detailed design
+
+We'll create an AST transform in [`packages/ember-template-compiler`](https://github.com/emberjs/ember.js/tree/master/packages/ember-template-compiler) which will emit a deprecation warning for all uses of `{{with}}`. The deprecation warning will be something like:
+
+```
+Using `{{with}}` is deprecated, please use a combination of `{{let}}` and `{{if}}` instead.
+```
+
+This new deprecation warning should link to some documentation which clearly describes when you should use `{{let}}` and when you should use `{{let}}` and `{{if}}` in combination.
+
+## How we teach this
+
+We'll mentiton the deprecation in an Ember point release blog post.
+
+The deprecation message will link to good documentation which describe why `{{with}}` was deprecated and gives clear guidelines on how to migrate to using either `{{let}}` or `{{let}}` and `{{it}}` in combination.
+
+## Drawbacks
+
+This adds a little churn.
+
+## Alternatives
+
+We could leave `{{with}}` as is. I don't believe that this is a good option as the name `{{with}}` is confusing.
+
+We could deprecate `{{with}}` and introduce `{{if-let}}` in Ember core. This RFC originally made that exact proposal, I was strongly persuaded of the [lack of need for `{{if-let}}` by @tcjr](https://github.com/emberjs/rfcs/pull/445#issuecomment-463594185).
+
+We could deprecate `{{with}}` and introduce `{{if-let}}` in an addon instead of Ember core.
+
+## Unresolved questions
+
+This [comment](https://github.com/emberjs/rfcs/issues/316#issuecomment-449489686) seems to indicate some support for `{{if-let}}` from the core team. Perhaps there are some compelling reasons for add it that are not currently included in this RFC?
```

Comment:
I _personally_ like `if-let`, but I think that this is just as easy to understand and doesn't rely on the reader to grok extra concepts....